Optimize sparse adam by fusing kernels

xiaomengy opened the pull request:
Summary: Optimize sparse adam by fusing kernels

Differential Revision: D16527697

soumith reviewed (changes requested):
`torch.optim` as a namespace and package has this really productive and important property that each file is self-contained, and is easily copy-pasteable. Several research works have done nothing but copy-paste `torch.optim` files, and then simply modify them to try out new algorithms.

I intend for this property to be preserved, because of how productive it makes people working on research in this area.

With that being said, if you want to write a fully fused optimizer that makes the code no longer accessible, but makes it considerably faster, I support your effort. The question then becomes "why not both", i.e. make it faster but also make it hackable.

Here are some ways to dig into and resolve this.

1. Describe what optimizations you are doing. Why is this code faster?
2. Is the algorithm decomposable into smaller chunks that can be fused separately and make the code readable? For example, what I mean is: can you transform the code into different `torch.xxx` building blogs like `torch.scatter` + `torch.matmul` + `torch.gather` (as an example) and still make it 80 to 90% as fast?
3. If (2) is not true, will something like `NestedTensor` keep the logic in Python but make it as fast (I know that'll be a future plan, as NestedTensor hasn't landed, but I want to get an idea of the future
4. Separately from (1) and (2), can you add a `fused=True` argument that takes the fast path (and is the default), but can you actually keep the slow path (add unit tests to make fused and unfused are never de-coupled)

Lastly, what are all of the other `optim` algorithms you are attempting to make fast? I didn't see this PR coming. Can you file Github issues for what you intend to do before-hand.

bddppq commented:
@soumith, re. your comments:

> 1. Describe what optimizations you are doing. Why is this code faster?

Sure, we will add some more explanations to the PR description. We will also attach numbers soon.

> 2. Is the algorithm decomposable into smaller chunks that can be fused separately and make the code readable? For example, what I mean is: can you transform the code into different torch.xxx building blogs like torch.scatter + torch.matmul + torch.gather (as an example) and still make it 80 to 90% as fast?

@BIT-silence and I have spent quite some time on discussing this option and we couldn't find a good way to achieve this without writing a custom fused kernel. (Happy to be proven wrong here.)

> 3. If (2) is not true, will something like NestedTensor keep the logic in Python but make it as fast (I know that'll be a future plan, as NestedTensor hasn't landed, but I want to get an idea of the future

AFAIK NestedTensor is not ready and doesn't have a clear committed timeline. We can sure start thinking about whether NestedTensor could be an alternative way, but it shouldn't be a blocker of this change.

> 4. Separately from (1) and (2), can you add a fused=True argument that takes the fast path (and is the default), but can you actually keep the slow path (add unit tests to make fused and unfused are never de-coupled)

I don't think this is a good idea, since a) this will unnecessarily increase the surface of optimizer api, we will need to add this new argument to *every* optimizer. b) to keep a more "hackable" code snippet that other researchers can more easily copy paste, a better place would probably be in pytorch/examples instead of pytorch core. Also I would like to discourage researchers to copy paste the implementation that have well known perf issue.

> Lastly, what are all of the other optim algorithms you are attempting to make fast? I didn't see this PR coming. Can you file Github issues for what you intend to do before-hand.

We are investigating adagrad.py, but atm not sure whether there are perf issue yet. Other than that we don't have any other planned changes for now. If there are non-trivial changes, we will definitely file PR and design discussions before hand. For changes like in this PR (that are non-api changing, fully backward compatible and purely convincing perf improvement) I would really hope we can move faster and do the discussion directly in a PR.

bddppq reviewed (commented):
@BIT-silence linter is complaining

```
./torch/optim/sparse_adam.py:3:1: F401 'math' imported but unused
```

bddppq commented on test/test_optim.py:
```diff
@@ -340,9 +356,17 @@ def test_sparse_adam(self):
+            test_cuda=False,
         )
-        with self.assertRaisesRegex(ValueError, "Invalid beta parameter at index 0: 1.0"):
+        self._test_rosenbrock_sparse(
```
You will need to skip this if cuda is not available

soumith commented:
@bddppq the hackability of the `optim` files is not negotiable, I want to preserve that property. If we are not doing with` fused=True`, we have to find a different mechanism, like the `jit`, but we do want to preserve this property.

zheng-xq commented:
I've talked with @bddppq offline. We agreed that both hackability and performance are important here. We have a few choices: 

1. The right solution down the road is use a graph compiler to generate the code. 
2. Before that happens, add a "fused" parameter. We can ignore this argument after switching to a compiler-based approach. 
    * We need to make sure the two paths are matching. Maybe a test to prove they produce the same results. 
3. An alternative is to remember the JIT hash of the current python code. If they match what we have today, it will call the faster fused C++ path. If not, it calls the slower Python code. This is a common practice among graph-based frameworks.

xiaomengy committed:
Optimize sparse adam by fusing kernels (#23522)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/23522

Optimize sparse adam by fusing kernels

Differential Revision: D16527697

fbshipit-source-id: e14ed0a6d94f772455dcd89c24a2264c355ad613

BIT-silence commented:
From our internal test, this PR can improve the SparseNN training QPS for 3-5x based on different models.

SsnL commented:
I don't have any saying in any of this, and parts of what I will say probably doesn't sound very nice. But I want to chime in here, because I really care about this project and sincerely hope PyTorch stays true to the open-source spirit, instead of becomes a FB repo that lives in a seemingly open-source shell. 

1. What @soumith asked for isn't even that much. Is it that hard to open an issue before working on an PR? It takes like 2 min max and I can't see any reason that it would make you move slower. You can even do this when waiting for PyTorch to build. 

2. Is it that hard to just write a 20 line self-contained script for benchmarking that will take like 5 min max? Saying 
    >From our internal test, this PR can improve the SparseNN training QPS for 3-5x based on different models. 

    means literally nothing for anyone from open-source, and probably very little even for FB people not knowing what your models are.

3. Why is this PR not requesting reviews from anyone who knows about PyTorch and `torch.optim` philosophies? I know in FB anyone can stamp anyone's diff, but that doesn't sound to me that anyone should freely change any part of any project as long as they have a friend.

4. Some of the reasonings don't make much sense to me either, and sound like they are made just to push this patch through. This makes me very worried. 

    Quote one comment above:

    > I don't think this is a good idea, since a) this will unnecessarily increase the surface of optimizer api, we will need to add this new argument to every optimizer. b) to keep a more "hackable" code snippet that researchers can more easily copy paste, a better place would probably be in pytorch/examples instead of pytorch core. Also I would like to discourage researchers to copy paste the implementation that have well known perf issue.

    1. > we will need to add this new argument to every optimizer.

        How is that even a concern if it is optional and hidden and not used by people for 99% of the times?

    2.  > a better place would probably be in pytorch/examples instead of pytorch core.

        So adding a flag for each optimizer is not a good idea, but writing slow versions of every pytorch optimizer in `pytorch/example` and not having anything to test that they do the correct thing is? 

    3. > discourage researchers to copy paste the implementation that have well known perf issue.

        Well-known? How? I wrote this optimizer and I am not aware of it. GitHub issue number? 

        Also, how is your proposal a better solution for copy-paste hacking? Moving the same unoptimized code to a different place that is untested and harder to find doesn't sound a move to the right direction. 

        Most ML researches don't even care much about speed.  

Finally, I am really glad to see things are getting sorted out in the correct direction, e.g., with the addition of `fused_kernel=True` flag. Thank you all for making the project better.

soumith commented on test/test_optim.py:
```diff
@@ -340,9 +356,32 @@ def test_sparse_adam(self):
+                sparse_only=True,
+                test_cuda=True,
+            )
+            self._test_rosenbrock_sparse(
```
@BIT-silence rosenbrock tests for fused=True and fused=False are not enough. Can you add tests to make sure fused=True and fused=False are computing the same function. This way, if we get a PR that modifies the fused=False codepath, the tests would fail if fused=True codepath doesn't get changed

dzhulgakov commented on torch/optim/sparse_adam.py:
```diff
@@ -64,41 +72,56 @@ def step(self, closure=None):
+                        torch._C._nn._sparse_adam_step(
+                            grad, exp_avg, exp_avg_sq, group['lr'], beta1,
+                            beta2, group['eps'], state['step'])
+                    exp_avg.add_(exp_avg_step)
```
if you make _sparse_adam_step modify the tensors in-place - would it be faster?

soumith commented:
@ssnl thanks for flagging, we can do better and we will do better. It's something that's happening over the distribution of PRs from our Facebook team (don't want to single out this PR as being specially different from the mean), and we need to fix ourselves over the distribution. It's not easy, because we are ramping up people more rapidly onto the project than perspectives and culture can get propagated -- but that's really no excuse.

@dzhulgakov @dwarakrajagopal @zheng-xq @jeffreyksmithjr @suo let's collectively get better at not alienating our core devs and work on structurally fixing this on the Facebook team.

@SsnL I hope this is not a one-off flagging, please continue to flag when you feel you don't get visibility or see PRs that shouldn't be, because it's going to continue to give the right signal for us.

dzhulgakov reviewed (commented):
Regarding process, docs and benchmarks, I think @soumith and @SsnL bring good points - we should and will do better. We should add a microbenchmarks for the ops we tweak (we can utilize new framework for it) and post result in the PR. It also helps tremendously with future optimizations.

For the actual implementation, fused_kernel flag seems pretty elegant. It'd take a longer time to get to JITing optimizers, though it's a great proof case for JIT

jeffreyksmithjr commented:
Thanks for your continued passion for our success as a project, @SsnL . It's important that we set a high bar for what it means to do open source right and continually stay accountable against that expectation. @soumith and @dzhulgakov are on top of steering us back in the right direction. I hope you'll continue to call out when we're falling short of the needs of the community. We really appreciate it.

gchanan commented on aten/src/ATen/native/cpu/optimizers_kernel.cpp:
```diff
@@ -0,0 +1,86 @@
+
+template <typename T>
+void SparseAdamStepKernelImpl(
+    T alpha,
```
why are these arguments template types on CPU but doubles on CUDA?  Comment?

gchanan commented on aten/src/ATen/native/cpu/optimizers_kernel.cpp:
```diff
@@ -0,0 +1,86 @@
+    T eps,
+    int64_t step,
+    TensorIterator* it) {
+  TORCH_CHECK(it->is_contiguous());
```
this isn't what TORCH_CHECK is for, see https://github.com/pytorch/pytorch/blob/3e0da2ab8e702e20abb3cbfbd71319aa71615568/c10/util/Exception.h#L169-L191.

Also, I couldn't quickly convince myself that all of the operands are actually guaranteed to be contiguous -- can you provide an argument why (in comments?).

fmassa commented:
@BIT-silence @bddppq another middle-ground option (as mentioned by @soumith) would be to use JIT to fuse the operations for us (specially if we are only fusing pointwise ops).

I gave it a quick try on a potential way of implementing this with the JIT, you can find it in https://gist.github.com/fmassa/240e61616146a4a254c8a546e8afd29d
The basic test that I wrote seems to work as expected.

Could you try benchmarking it to see if there are any significant slowdowns compared to the hand-fused kernel?

I would expect it to be (slightly) slower and use more memory if we don't use the JIT compared to the original python version, because I removed the in-place ops.

But if we always use the JIT (which could be a possibility, given that the JIT is a first-class citizen in PyTorch), then it should not be a problem.

I inspected the generated kernel for `update_step ` and it looks like it is entirely fused by the compiler:
```
graph(%grad_values.1 : Float(*),
      %old_exp_avg_values.1 : Float(*),
      %old_exp_avg_sq_values.1 : Float(*),
      %beta1.1 : float,
      %beta2.1 : float,
      %eps.1 : float,
      %step_size.1 : float):
  %7 : int = prim::Constant[value=1]() # sparse_adam.py:177:67
  %10 : float = aten::sub(%7, %beta1.1) # sparse_adam.py:177:67
  %14 : float = aten::sub(%7, %beta2.1) # sparse_adam.py:178:78
  %20 : float = aten::neg(%step_size.1) # sparse_adam.py:184:12
  %fact.1 : Float(*), %exp_avg_sq_update_values.1 : Float(*), %exp_avg_update_values.1 : Float(*) = prim::FusionGroup_0(%old_exp_avg_sq_values.1, %old_exp_avg_values.1, %grad_values.1, %20, %eps.1, %14, %10)
  %23 : (Float(*), Float(*), Float(*)) = prim::TupleConstruct(%exp_avg_update_values.1, %exp_avg_sq_update_values.1, %fact.1)
  return (%23)
with prim::FusionGroup_0 = graph(%13 : Float(*),
      %17 : Float(*),
      %26 : Float(*),
      %4 : float,
      %7 : float,
      %21 : float,
      %30 : float):
  %27 : int = prim::Constant[value=2]() # sparse_adam.py:178:48
  %32 : int = prim::Constant[value=1]() # sparse_adam.py:177:67
  %33 : Float(*) = aten::sub(%26, %17, %32) # sparse_adam.py:177:30
  %exp_avg_update_values.1 : Float(*) = aten::mul(%33, %30) # sparse_adam.py:177:30
  %28 : Float(*) = aten::pow(%26, %27) # sparse_adam.py:178:33
  %25 : Float(*) = aten::sub(%28, %13, %32) # sparse_adam.py:178:33
  %exp_avg_sq_update_values.1 : Float(*) = aten::mul(%25, %21) # sparse_adam.py:178:33
  %numer.1 : Float(*) = aten::add(%exp_avg_update_values.1, %17, %32) # sparse_adam.py:180:13
  %oo.1 : Float(*) = aten::add(%exp_avg_sq_update_values.1, %13, %32) # sparse_adam.py:181:10
  %11 : Float(*) = aten::sqrt(%oo.1) # sparse_adam.py:182:13
  %denom.1 : Float(*) = aten::add(%11, %7, %32) # sparse_adam.py:182:13
  %5 : Float(*) = aten::mul(%numer.1, %4) # <string>:2:10
  %fact.1 : Float(*) = aten::div(%5, %denom.1) # sparse_adam.py:184:12
  return (%fact.1, %exp_avg_sq_update_values.1, %exp_avg_update_values.1)
```
